### PR TITLE
proof-producer: human-readable assignment table output

### PR DESCRIPTION
**Pull request by oclaw**
Moved output_artifacts library from `zkevm-framework` to `proof-producer` and enabled human-readable table writing as an option for binary

Detailed changes:
- moved `export_table` impl from `blueprint` to `zk` `crypto3` library
- refactored `Range(s)` classes a bit
- bound `OutputArtifacts&` to `boost program options` in a modern way
- added full table write mode (and full sub-option for rows each column type ranges)
- added write calls to `assignment_table_writer` and `Prover` classes
- added optional table write to all prover stages that include work with table

### CODE REVIEW

File: proof-producer/bin/proof-producer/src/arg_parser.cpp [hunk 1]
```diff
@@ -117,6 +118,8 @@ namespace nil {
                  "Files containing polynomials combined-Q, 1 per prover instance.")
                 ("proof-of-work-file", make_defaulted_option(prover_options.proof_of_work_output_file), "File with proof of work.");
 
+            nil::proof_generator::register_output_artifacts_cli_args(prover_options.output_artifacts, config);
```
[the remark this answers is not in the document]
Comment (oclaw) [reply proof-producer/bin/proof-producer/src/arg_parser.cpp]: fixed

File: proof-producer/libs/output_artifacts/src/output_artifacts.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                }
+                result.push_back(maybe_ranges.value());
+                v = boost::any(result);
+            }
+        }
+
+        void register_output_artifacts_cli_args(OutputArtifacts& to_fill, po::options_description& cli_options) {
+            
+                cli_options.add_options()
+                ("assignment-table-output-file", make_defaulted_option(to_fill.output_filename),
```
[the remark this answers is not in the document]
Comment (oclaw) [reply proof-producer/libs/output_artifacts/src/output_artifacts.cpp]: changed option names to debug

File: proof-producer/libs/output_artifacts/src/output_artifacts.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                 "Output filename for assignment tables. If omitted or set to '-', write to stdout")
```
[the remark this answers is not in the document]
Comment (oclaw) [reply proof-producer/libs/output_artifacts/src/output_artifacts.cpp]: done

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp [hunk 4]
```diff
@@ -627,6 +629,52 @@ namespace nil {
                 return true;
             }
 
+            bool save_text_assignment_table_to_file(const OutputArtifacts& opts) {
```
[the remark this answers is not in the document]
Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp]: done

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/arithmetization/plonk/export.hpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+
+#ifndef CRYPTO3_ZK_PLONK_PLACEHOLDER_TABLE_EXPORT_HPP
+#define CRYPTO3_ZK_PLONK_PLACEHOLDER_TABLE_EXPORT_HPP
+
+#include <ostream>
+#include <vector>
+
+#include <nil/crypto3/zk/snark/arithmetization/plonk/assignment.hpp>
+#include <nil/crypto3/zk/snark/arithmetization/plonk/table_description.hpp>
```
[the remark this answers is not in the document]
Comment (oclaw) [reply crypto3/libs/zk/include/nil/crypto3/zk/snark/arithmetization/plonk/export.hpp]: yep, removed

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp [hunk 6]
```diff
@@ -636,6 +638,52 @@ namespace nil {
...
+                BOOST_LOG_TRIVIAL(debug) << "Public input columns to print: "
+                                         << opts.public_input_columns.to_string();
+                BOOST_LOG_TRIVIAL(debug) << "Constant columns to print: "
+                                         << opts.constant_columns.to_string();
+                BOOST_LOG_TRIVIAL(debug) << "Selector columns to print: "
+                                         << opts.selector_columns.to_string();
+
+
+                const auto write = [&](std::ostream& out) -> bool {
+                    return nil::proof_generator::assignment_table_writer<Endianness, BlueprintField>::write_text_assignment(
```
Comment (akokoshn) [inline proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp:664]: useless `nil::proof_generator`

**Review by akokoshn — APPROVED** [review]

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp (hunk 6 above)
Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp:664]: will fix in the next PRs

**Pull request merged**